server/src/main/java/org/apache/druid/server/ClientQuerySegmentWalker.java

```diff
@@ -431,6 +448,101 @@ private DataSource inlineIfNecessary(
         );
   }
 
+  /**
+   * This method returns the datasource by populating all the {@link QueryDataSource} with correct nesting level and
+   * sibling order of all the subqueries that are present.
+   * It also plumbs parent query's id and sql id in case the subqueries don't have it set by default
+   *
+   * @param dataSource       Datasource whose subqueries need to be populated
+   * @param parentQueryId    Parent Query's ID, can be null if do not need to update this in the subqueries
+   * @param parentSqlQueryId Parent Query's SQL Query ID, can be null if do not need to update this in the subqueries
+   * @return DataSource populated with the subqueries
+   */
+  private DataSource generateSubqueryIds(
+      DataSource dataSource,
+      @Nullable final String parentQueryId,
+      @Nullable final String parentSqlQueryId
+  )
+  {
+    Queue<DataSource> queue = new LinkedList<>();
+    queue.add(dataSource);
+
+    /*
+    Performs BFS on the datasource tree to find the nesting level, and the sibling order of the query datasource
+     */
+    Map<DataSource, Pair<Integer, Integer>> queryDataSourceToSubqueryIds = new HashMap<>();
+    int level = 1;
+    while (!queue.isEmpty()) {
+      int size = queue.size();
+      int siblingOrder = 1;
+      for (int i = 0; i < size; ++i) {
+        DataSource currentDataSource = queue.poll();
+        if (currentDataSource instanceof QueryDataSource) {
+          queryDataSourceToSubqueryIds.put(currentDataSource, new Pair<>(level, siblingOrder));
```

Comment:
It is not clear to me why we are not calling `insertSubQueryId` here itself. That is we have the level information already so we can populate the ids for the query corresponding to this QueryDataSource here itself instead of saving it inside a map and doing it later. am I missing something?

Reply:
The `Query#withSubqueryId()` method returns a new subquery with modified context, as does `Datasource#withChildren`. In order to actually modify the subquery id, we would have to go in a bottom's up manner (from the innermost datasource) and replace the child of the parent datasource, and then return this new parent datasource to be replaced in it's own parent datasource and so forth. I failed to come up with a way of doing that in a loop.

Reply:
I see. Thank you for explaining that.